Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp

```diff
@@ -0,0 +1,1069 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WasmIPIntGenerator.h"
+
+#if ENABLE(WEBASSEMBLY)
+
+#include "BytecodeGeneratorBaseInlines.h"
+#include "BytecodeStructs.h"
+#include "InstructionStream.h"
+#include "JSCJSValueInlines.h"
+#include "Label.h"
+#include "WasmCallingConvention.h"
+#include "WasmContext.h"
+#include "WasmFunctionIPIntMetadataGenerator.h"
+#include "WasmFunctionParser.h"
+#include "WasmGeneratorTraits.h"
+#include <variant>
+#include <wtf/CompletionHandler.h>
+#include <wtf/RefPtr.h>
+
+/* 
+ * WebAssembly in-place interpreter metadata generator
+ * 
+ * docs by Daniel Liu <daniel_liu4@apple.com / danlliu@umich.edu>; 2023 intern project
+ * 
+ * 1. Why Metadata?
+ * ----------------
+ * 
+ * WebAssembly's bytecode format isn't always the easiest to interpret by itself: jumps would require parsing
+ * through many bytes to find their target, constants are stored in LEB128, and a myriad of other reasons.
+ * For IPInt, we design metadata to act as "supporting information" for the interpreter, allowing it to quickly
+ * find important values such as constants, indices, and branch targets.
+ * 
+ * 2. Metadata Structure
+ * ---------------------
+ * 
+ * Metadata is kept in a vector of UInt8 (bytes). We handle metadata in "metadata entries", which are groups of
+ * 8 metadata bytes. We keep metadata aligned to 8B to improve access times. Sometimes, this results in higher
+ * memory overhead; however, these cases are relatively sparse. Each instruction pushes a certain number of
+ * entries to the metadata vector.
+ *
+ * 3. Metadata for Instructions
+ * ----------------------------
+ *
+ * Instructions not listed here do not utilize any metadata.
```

Comment:
We may want to re-think this comment. If it is too specific, and someone forgets to update it, it might get very confusing later on

Reply:
Makes sense, yeah. Currently using it as a centralized location to keep track of all the metadata (as opposed to digging around in the code). I'll make an update to specifically identify the instructions that don't have metadata.